### PR TITLE
module 3 completed

### PR DESCRIPTION
**Pull request by anastasii27**

### CODE REVIEW

File: common/src/main/java/com/epam/esm/config/SpringConfig.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;
+import javax.sql.DataSource;
+
+@ComponentScan(basePackages = {"com"})
+@Configuration
+public class SpringConfig implements WebMvcConfigurer {
+
+    @Bean
+    @ConfigurationProperties(prefix = "spring.datasource")
+    public DataSource dataSource() {
```
Comment (aolshevskiy) [inline common/src/main/java/com/epam/esm/config/SpringConfig.java:23]: This not needed with Spring Boot.

File: common/src/main/java/com/epam/esm/config/SpringConfig.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      return DataSourceBuilder.create().build();
+    }
+
+    @Bean
+    public ModelMapper modelMapper() {
+        return new ModelMapper();
+    }
+
+    @Bean
+    public LocalSessionFactoryBean sessionFactory() {
```
Comment (aolshevskiy) [inline common/src/main/java/com/epam/esm/config/SpringConfig.java:33]: Not needed.

File: common/src/main/java/com/epam/esm/config/SpringConfig.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    @Profile("test")
+    public EmbeddedDatabase embeddedDatabase(){
+        return new EmbeddedDatabaseBuilder()
+                .setType(EmbeddedDatabaseType.H2)
+                .build();
+    }
+
+    @Bean
+    @Profile("test")
+    public PlatformTransactionManager transactionManager() {
```
Comment (aolshevskiy) [inline common/src/main/java/com/epam/esm/config/SpringConfig.java:51]: Not needed.

File: common/src/main/java/com/epam/esm/model/Certificate.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+public class Certificate {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private long id;
+    private String name;
+    private String description;
+    private BigDecimal price;
+    private LocalDateTime createDate;
+    @Convert(converter = ZoneIdConverter.class)
+    private ZoneId createDateTimezone;
```
Comment (aolshevskiy) [inline common/src/main/java/com/epam/esm/model/Certificate.java:28]: It's very rare requirement to store timezone in the database. Usually UTC is used except for special cases like time of some event that happens in some special timezoned context. Please reconsider if you really need timezone for your datetime types.

File: common/src/main/java/com/epam/esm/model/Certificate.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    private LocalDateTime lastUpdateDate;
+    @Convert(converter = ZoneIdConverter.class)
+    private ZoneId lastUpdateDateTimezone;
+    private long duration;
+    @ManyToMany(cascade = {CascadeType.MERGE, CascadeType.PERSIST})
+    @JoinTable(name = "tag_m2m_gift_certificate",
+               joinColumns = { @JoinColumn(name = "giftCertificateId")},
```
Comment (aolshevskiy) [inline common/src/main/java/com/epam/esm/model/Certificate.java:35]: Current consesus expects uniform naming convention for both column names and table names. Also current convention is to use snake case for the database. Spring Boot has such options.

File: common/src/main/java/com/epam/esm/model/Order.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private long id;
+    private LocalDateTime purchaseDate;
+    @Convert(converter = ZoneIdConverter.class)
+    private ZoneId purchaseTimeZone;
+    private BigDecimal cost;
+    private long userId;
+    @ManyToMany(fetch = FetchType.LAZY, mappedBy = "orders")
+    @NotAudited
```
Comment (aolshevskiy) [inline common/src/main/java/com/epam/esm/model/Order.java:32]: Why do we need Envers here?

File: giftcertificate/src/main/java/com/epam/esm/dto/CertificateDto.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
+package com.epam.esm.dto;
+
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonPOJOBuilder;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import org.springframework.hateoas.RepresentationModel;
```
Comment (aolshevskiy) [inline giftcertificate/src/main/java/com/epam/esm/dto/CertificateDto.java:9]: This import is not used.

File: giftcertificate/src/main/java/com/epam/esm/repository/CertificateRepositoryImpl.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+import java.util.Map;
+import java.util.Optional;
+
+@Repository("certificateRepository")
+public class CertificateRepositoryImpl implements CertificateRepository {
+    private static final String ALL_CERTIFICATES = "SELECT c FROM Certificate c";
+    private static final String GET_CERTIFICATE = "SELECT c FROM Certificate c WHERE c.name=?1 AND c.description=?2";
+    private static final String GET_USER_CERTIFICATES = "SELECT c FROM Certificate c JOIN c.orders o WHERE o.userId=?1";
+    @Autowired
+    private SessionFactory sessionFactory;
```
Comment (aolshevskiy) [inline giftcertificate/src/main/java/com/epam/esm/repository/CertificateRepositoryImpl.java:19]: Please replace with EntityManager.

File: giftcertificate/src/main/java/com/epam/esm/repository/SelectFilteredQuery.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        return FilterParam.stream()
+                .map(e->e.name().toLowerCase())
+                .filter(mapKey::contains)
+                .findFirst().orElse(SPACE);
+    }
+
+    private String getWhereConditionForSearchParam(String value, String paramName){
+        return new StringBuilder().append(valueOf(paramName.toUpperCase()).getParam())
+                .append(LIKE)
+                .append(value)
```
Comment (aolshevskiy) [inline giftcertificate/src/main/java/com/epam/esm/repository/SelectFilteredQuery.java:109]: This code is vulnerable to SQL injection. Please use SQL parameters.

File: user/src/main/java/com/epam/esm/repository/UserRepositoryImpl.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+import org.hibernate.SessionFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Repository;
+import java.util.*;
+
+@Repository("userRepository")
+public class UserRepositoryImpl implements UserRepository {
+    private static final String USERS_LIST  = "SELECT u FROM User u";
+    @Autowired
+    private SessionFactory sessionFactory;
```
Comment (aolshevskiy) [inline user/src/main/java/com/epam/esm/repository/UserRepositoryImpl.java:15]: Please replace with EntityManager.

**Review by aolshevskiy — CHANGES_REQUESTED** [review]
General notes:
1. Usage of hibernate specific features is prohibited per requirements. Please use EntityManager and other JPA interfaces only.
2. Root pom.xml is not under Git
Non conflicting requirements hold from previous modules:
1. Layers should be separate from each other. Not modules for entities
2. No unit tests just integration once again

File: giftcertificate/src/test/java/com/epam/esm/service/CertificateServiceImplTest.java [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+    @Test
+    void update_whenCertificateExists_thenReturnUpdatedCertificate() throws Exception {
+        certificate.setTags(Collections.emptySet());
+        certificateDto.setTags(Collections.emptyList());
+        when(certificateRepository.findById(1)).thenReturn(Optional.of(certificate));
+        when(updatedCertificate.getUpdatedCertificate(certificateDto, certificateDto)).thenReturn(certificateDto);
+        when(certificateRepository.update(certificate)).thenReturn(Optional.of(certificate));
+        PowerMockito.doNothing()
+                .when(tagService, "updateCertificateTags", certificateDto.getTags(), certificateDto.getId());
+
```
Comment (natashlapakova) [inline giftcertificate/src/test/java/com/epam/esm/service/CertificateServiceImplTest.java:139]: Mockito.doNothing() can be used instead

File: common/src/main/java/com/epam/esm/audit/listener/CertificateListener.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+    @PreRemove
+    public void preRemove(Certificate certificate){
+        insertIntoAuditTable(DELETE, certificate);
+    }
+
+    private void insertIntoAuditTable(AuditAction action, Certificate certificate){
+        entityManager.persist(new CertificateHistory(0, certificate.getId(), action, certificate.getName(),
+                certificate.getDescription(), certificate.getPrice(), certificate.getCreateDate(), certificate.getCreateDateTimezone(),
+                certificate.getLastUpdateDate(), certificate.getLastUpdateDateTimezone(), certificate.getDuration()));
```
Comment (dzmitrySazonov) [inline common/src/main/java/com/epam/esm/audit/listener/CertificateListener.java:34]: better to replace such multi-args constructors with builders

**Pull request merged**